### ISSUE TITLE
Avoid page navigation during critical css generation (causes errors)

### ISSUE THREAD

**Issue by hummal** [pocketjoso/penthouse#202]
For some URLs Iam checking Iam getting an strange error, which is properbly puppeteer releated. Think of updating to a newer version.

Message was:
`{ Error: Error: Protocol error (Runtime.callFunctionOn): Execution context was destroyed. undefined
    at /var/www/bamboo_deploy/1190/vendor/tcms/atf-optimizer/node_modules/penthouse/lib/index.js:269:23
    at Generator.next (<anonymous>)
    at step (/var/www/bamboo_deploy/1190/vendor/tcms/atf-optimizer/node_modules/penthouse/lib/index.js:45:191)
    at /var/www/bamboo_deploy/1190/vendor/tcms/atf-optimizer/node_modules/penthouse/lib/index.js:45:361
    at process._tickCallback (internal/process/next_tick.js:109:7)
  stderr: 'Error: Protocol error (Runtime.callFunctionOn): Execution context was destroyed. undefined' }
`

Maybe this is related: https://www.bountysource.com/issues/51332369-when-would-an-error-cannot-find-context-with-specified-id-undefined-happen

Also I want to mention that we need to install a lot of dependeciens with puppeteer. So I will have 3 new users in my passwd file. Keeps me feeling bad.

The problem with the failing id is much more important for me.

**Comment by pocketjoso** [comment]
Based on that discussion, it should be fixed in the next puppeteer release (0.14) - so let's retry then.

**Comment by pocketjoso** [comment]
I know puppeteer install can be quite painful, hopefully that will be improved in the library itself. If you really want to you could revert to `0.11.13`, the version that still used phantomjs. It's no longer maintained however.

**Comment by hummal** [comment]
I hope so. There is a little difference in my error message and the one in the thread. Its about execution context was destroyed and not an id is missing. 

Returning to an older version is not an option. Iam not into using software without updates. :)

**Comment by hummal** [comment]
@pocketjoso what do you expect as output from debuglog function?

Code like this: 
```javascript
 stdErr += debuglog('call generateCriticalCssWrapped');
```
results in _stdErr = undefinedundefinedundefined_

**Comment by hummal** [comment]
Ok I solved the problem by simply adding a return ""; to the end of the debuglog function.
PR incoming

**Comment by hummal** [comment]
@pocketjoso I opend an issue at puppeteer and they had some interesting information. The result would be that you should include these handlings because it could happen to anyone.

https://github.com/GoogleChrome/puppeteer/issues/1591#issuecomment-355245777

**Comment by pocketjoso** [comment]
@hummal thanks for the digging - indeed it would be a good feature to add to Penthouse to lock the page from navigating during the critical css generation, to avoid these errors. If you're interested feel free to work on a PR, otherwise as usual I'll get around to it eventually. Please let me know here if you are going to start such work so we can avoid duplicate work!

**Comment by hummal** [comment]
I will let you know when I start working.

**Comment by hummal** [comment]
So I will start now to check if we can fix this issue

**Comment by pocketjoso** [comment]
Great, thanks, I will leave it to you then and review your PR when it comes.

**Comment by hummal** [comment]
After some digging it seems we have an race condition  with pageLoadSkipTimeout and the successfull return of the response.

Here is my setup:
```javascript
  timeout:                 90000,
  pageLoadSkipTimeout:     10000,
  renderWaitTime:          200,
  blockJSRequests:         true,    
```

I tried some 1000 tests. with one url of our own servers.
nginx access log was helping me.
- every time penthouse finished successfully was when the page responded under 10 seconds
- every time the response took more than 10s penthouse didn't wait any longer for the response and the client quit the connection with status 499, although we had a timeout set of 90s
- in a small chance of having the response in 10 to 10.2 seconds there is a race condition where the client gets the status 200 and while the resonse from the server is incoming we decide to cut the connection.

Here are the logs:
`
// Response in < 10s | status: 200 | bytes sent: 501439 | time till response: 7.367
[31/Jan/2018:17:23:37 +0100] 10.111.26.101 - - - https 200 501439 7.367

// Response in > 10.2 | status: 499 (Client disconnected)  | bytes sent: 0 | time till response: 10.960 ( due to abort)
[31/Jan/2018:16:47:15 +0100] 10.111.26.101 - - - https 499 0 10.960

// Response in  10s - 10.2s | status: 200 | bytes sent: 490601 | time till response: 10.086
[31/Jan/2018:17:24:43 +0100] 10.111.26.101 - - - https 200 490601 10.086
`

The important thing is that we got a status 200 but not the whole response data was transfered because we aborted the connection before.

Next step is to check why we are doing this. Thinking of the strange sleep function in purneNonCriticalSelectors.js. Maybe there is a race condition.

**Comment by pocketjoso** [comment]
Interesting.
Check the [pageLoadSkipTimeout logic](https://github.com/pocketjoso/penthouse/blob/master/src/core.js#L117-L142), if you didn't already. It doesn't reject the page loading (shouldn't quit the connection), but we do stop waiting for it after the time specified, and move on to start extracting the critical css.

For your sake, what happens if you increase the `pageLoadSkipTimeout` to a really high value, does that make this error go away in your setup?

**Comment by hummal** [comment]
Iam on it. 👍 

If I increase the pageLoadSkipTimeout to 20000 it never happened in my tests.

**Comment by hummal** [comment]
Here we go

```
penthouse:core page load waiting ABORTED after 10.1s.  +10s
  penthouse:core page load DONE +1ms
  penthouse:core build selector profile +1ms
[0201/104704.874645:INFO:CONSOLE(5)] "debug: pruneNonCriticalSelectors", source:  (5)
  penthouse:core pruneNonCriticalSelectors +10ms
  penthouse:core cleanupAndExit +11ms
  penthouse remove browser page for generateCriticalCss after ERROR, now: 0 +10s
  penthouse closed browser +9ms
Error: Protocol error (Runtime.callFunctionOn): Execution context was destroyed. undefined
```

**Comment by pocketjoso** [comment]
Cannot reproduce. I lowered pageLoadSkipTimeout to `2500` and ran penthouse with two slow domains, works as expected.

Please try your own setup with some other page urls, that you don't own. Do you still get the same problem?

Are you on the latest penthouse version/master?

**Comment by hummal** [comment]
This only happens in a very short time frame. I have an domain which runs on a cache but when I refresh the cache it could happen that it needs up to 10s for pageload. I need to run this workflow up to 50 times and then I get the hit. 

Problem as you can see above is that the pruneNonCriticalSelectors script, which is injected in the puppeteer page runs into an error because the page is not loaded completely.

'''
await Promise.race([
          loadPagePromise,
          new Promise(resolve => {
            // instead we manually _abort_ page load after X time,
            // in order to deal with spammy pages that keep sending non-critical requests
            // (tracking etc), which would otherwise never load.
            // With JS disabled it just shouldn't take that many seconds to load what's needed
            // for critical viewport.
            setTimeout(() => {
              if (waitingForPageLoad) {
                debuglog(
                  'page load waiting ABORTED after ' +
                    pageLoadSkipTimeout / 1000 +
                    's. '
                )
                resolve()
              }
            }, pageLoadSkipTimeout)
          })
        ])
'''
Here is where the problem seems to come from, but I can't say why. 
I replaced the Promise.race logic with Promise.all and used rejects to determine when one is finished. 
Also I injected the pageload INTO the page with evaluate. 
The reason is: 
- pageLoadSkipTimeout should start with the response of the document because otherwise it would do the same as timeout
- when started after document (the first request/response) it is a real counter for "what comes after" and can be more precise
- if evaluated in browser page it can trigger window.stop() which stops all further requests

Sideeffects: injecting and evaluating the time consts some time (~100ms) but I guess this is ok due to the fact that it seems to be more natural.

Im in testing now to see if that fixed my problem

**Comment by hummal** [comment]
So Iam finished and the fix is comfirmed working.

Iam getting 
```
 https 200 490601 7.970
```
so it means the page is not loaded completely. This is due to my very short pageLoadSkipTimeout frame.
In the previous version we got the problem, that it broke the process and there was no css generated.

In my fix it worked as shown below:
´´´bash
  penthouse:core new page opened in browser +73ms
  penthouse:core viewport set +3ms
  penthouse:core blocking js requests +2ms
  penthouse:core page load start +0ms
  penthouse:core RESPONSE URL: xxx +8s
  penthouse:core pageLoadSkipTimeout injected on dom creation +0ms
  penthouse:core pageLoadSkipTimeout [10] +25ms
  penthouse:core pageLoadSkipTimeout - page load waiting ABORTED after 0.01s.  +15ms
  penthouse:core RACE RESULT:  pageLoadSkipTimeout +0ms
  penthouse:core page load DONE +1s
  penthouse:core build selector profile +66ms
  penthouse:core pruneNonCriticalSelectors +30ms
  penthouse:core waited for renderWaitTime: 500 +509ms
  penthouse:core filterSelectors BEFORE +0ms
  penthouse:core filterSelectors AFTER +92ms
  penthouse:core pruneNonCriticalSelectors done, now cleanup AST +5ms
´´´
 so no breaking anymore. Will serve a pull request asap

**Comment by hummal** [comment]
So after a real test with our companys setup it seems the problem is still there. 
The purpose of injecting the pageLoadTimeout script was to being able to do window.stop(), which was a try to prevent the page from doing wrong. 
Iam not sure of the `pruneNonCriticalSelectors `script leads to the page crash. But the error comes from page.evaluate. 
```bash
penthouse:core page.evaluate - ERROR Error: Protocol error (Runtime.callFunctionOn): Cannot find context with specified id undefined
```
And you are right. I will try to find out why it is crashing.

**Comment by hummal** [comment]
To get deeper into the code I have some questions:
Iam using a child_process to launch penthouse multiple times at once. At the moment 1 per core.
`let worker_process = child_process.fork(path.join(workerLib), options, processOptions);`
I found the logic for browserPagesOpen and Iam not sure when there will be more than 1 page opened in one browser. 
How could this happen when you only allow one url in a penthouse call?

To get deeper in our problem right now. When using the debug logs I found that the first browser crashes after the first cleanupAndExit script is running.
```
penthouse:core page load DONE +0ms
  penthouse:core build selector profile +41ms
  penthouse:core pruneNonCriticalSelectors +25ms
  penthouse:core cleanupAndExit start +381ms
  penthouse:core cleanupAndExit end +1ms
  penthouse remove browser page for generateCriticalCss after ERROR, now: 0 +8s
Chromium unexpecedly not opened - crashed?
_browserPagesOpen: 1
url: https://....
AST children: 5099
restarting chrome after crash
  penthouse no browser instance, launching new browser.. +6ms
```
This makes no sense to me because when the browser is a child process how could the cleanup script of child process 1 kill the browser of child process 2? To be clear: Iam not sure that this is really happening. Just discussing

**Comment by hummal** [comment]
Ok after I adjusted my test directly in penthouse I can confirm that the reason for the error
`page.evaluate - ERROR Error: Protocol error (Runtime.callFunctionOn): Cannot find context with specified id undefined`
is related to the child process logic and penthouse.

To reproduce use this test:
- create a folder "tmp" 
- create a file test.mjs with following content:
```javascript
import childProcess from 'child_process'

const workerToStart = 20 // this is the trick > 5 is ok in my case, 10 sometimes, but 20 everytime failes
const url = 'https://jonassebastianohlsson.com/criticalpathcssgenerator/'

for (let i = 0; i < workerToStart; i++) {

  console.log("WORKER STARTED %s", i)

  let workerProcess = childProcess.fork('./tmp/pmodule.mjs', [url, i])

  workerProcess.on('error', err => console.error("WORKER ERROR", err))

  workerProcess.on('close', (code) => {
    console.log("WORKER ENDED %s", i)
  })
}
```
- create a file called pmodule,mjs 
```javascript
import penthouse from '../lib'

const url = process.argv[2]
const i = process.argv[3]

penthouse({
  url:       url,       // can also use file:/// protocol for local files
  cssString: 'body { color; red }', // the original css to extract critcial css from
 
  // OPTIONAL params
  width:                   1300,                    // viewport width
  height:                  900,                    // viewport height
  keepLargerMediaQueries:  true,  // when true, will not filter out larger media queries
  timeout:                 60000,                 // ms; abort critical CSS generation after this timeout
  pageLoadSkipTimeout:     5000,         // ms; stop waiting for page load after this timeout (for sites with broken page load event timings)
  maxEmbeddedBase64Length: 1000,  // characters; strip out inline base64 encoded resources larger than this
  userAgent:               'Penthouse Critical Path CSS Generator', // specify which user agent string when loading the page
  renderWaitTime:          500,            // ms; render wait timeout before CSS processing starts (default: 100)
  blockJSRequests:         true,          // set to false to load (external) JS (default: true)
  strict:                  false,                  // set to true to throw on CSS errors
  puppeteer:               {
    getBrowser: undefined        // A function that resolves with a puppeteer browser to use instead of launching a new browser session
  }
}).then(criticalCss => {
  console.log("FINISHED")
  process.exit()
}).catch(err => {
  // handle the error
  console.log(err)
  process.exit()
})
```
- run
`DEBUG="penthouse*,-penthouse:css-cleanup*,-penthouse:preformatting*" node --experimental-modules tmp/test.mjs`

**Comment by pocketjoso** [comment]
Thanks for looking closely into this. I will take a closer look later, but note in general (not document enough I guess) regarding parallel processing and penthouse:
* by default, if you call penthouse again while another call is still in process, penthouse will re-use the same chromium (puppeteer) browser, and just open a new tab instead. This is a bit faster than open and closing browsers.
* if you pass the experimental (and intentionally not documented) `unstableKeepBrowserAlive` param, the launched browser will always be re-used (new tabs opened) and the browser will never be closed by Penthouse. Useful only if you are running a long-running node process that you will later kill manually (which will close the browsers)
* you can also launch a puppeteer browser yourself (see docs but basically `puppeteer.launch().then((browser)`), in which case behavior will be the same as above - new tabs, and `Penthouse` will never kill the browser - it is now your responsibility.

This way you can get much more efficient critical css generation - however how many cores it will utilise is up to your node configuration.

---

I have never looked into spawning child processes for Penthouse calls, so I can't say anything yet about how it interacts with above parallelization logic. I am however running penthouse effectively in large scale just relying on the parellization, without child processes, and it works well for me.

**Comment by hummal** [comment]
Thanks for the details. Unfortunately I can not see that penthouse will use the other browser. Thats why Iam asking. You can reproduce this by using the test above. As for the debug logs it shows that there is always a new browser. That was the reason why I was asking.

To the problem:
```javascript
      await page.waitFor(100)
      const criticalSelectors = await page.evaluate(pruneNonCriticalSelectors, {
        selectors,
        renderWaitTime
      })
```
With this `await page.waitFor(100)`it runs flawlessly through all my files and urls. Only the browser was crashing sometimes but you recovery logic works as expected. 
Also my test above with 20 workers is now running without any error. 
So the conclusion is that this is indeed related to puppeteer and not a race condition in penthouse.

Here is the task I was reading about a similar problem
https://github.com/GoogleChrome/puppeteer/issues/1325#issuecomment-357645094

**Comment by hummal** [comment]
I fixed 2 missing error rejects and it now restarts on execution context losing like you wanted it to be. 
The execution context problem is nearly gone but this is a real puppeteer issue and also not fixed in 1.0.
With the current status it is working as it should and I didn't lost a single page request because of the restarting after error. 
Having 8 parallel penthouse process running with each at least 5 urls to check and 24 different files to process. Had 3 times crashed browser (context lost) and this was NOT due to navigating away. 
The main impact was `page.waitFor(500)` which leads to smoother running.
Check the pull request and tell me if you are not happy with anything.

By the way: I really like your effort in writing penthouse but I need to say that in my opinion the code is ulgy. In its functional way it is very hard to read. What do you thing of refactoring to es6 classes and less event bouncing ( Promises )?

**Comment by pocketjoso** [comment]
@hummal will have to take time and give you a response later, but just quickly - I don't see any pull request, did you miss to send it?

**Comment by hummal** [comment]
#227 here you go.
I was just writing the comment :smiley:

**Comment by pocketjoso** [comment]
I reviewed the PR now, and to respond to your last comment:

> By the way: I really like your effort in writing penthouse but I need to say that in my opinion the code is ulgy. In its functional way it is very hard to read. What do you thing of refactoring to es6 classes and less event bouncing ( Promises )?

The code can surely be improved as it has mostly stayed in the same format since the beginning of the repo, when I was quite new to JS - but I am not sold on switching to Class usage. Making the code clean is not the only goal I have in mind, I also try to make Penthouse as fast as possible, which sometimes does lead to more complicated code, see #224 f.e. where I'm trying some increased parallelisation. Regarding promises (and async/await), I don't see how you mean that would change with a move to classes (async code would remain async code)?

**Comment by pocketjoso** [comment]
Reflecting a bit on what's going on here too, I want to say that:
- regardless of everything else - patching up the error handling in these _edge_ cases in Penthouse is a good thing, so thanks for working on this with me!
- I wonder however if you would really come across these errors at all if you used penthouse in a more standard way.

To clarify: in your test example you are starting up 20 chromium browsers (and page loads) at once. It is not unlikely that you run out of memory or even that your CPU becomes super busy, causing all the issues you see (that don't happen normally). Can you try running in a more recommended way, see below (I will put these examples in the README tomorrow):

```js
import puppeteer from 'puppeteer'
import penthouse from '../lib'

const url = process.argv[2]

// 20 here still means 20 tabs at the same time (but in the _same_ browser);
// my recommendation would be to cap at 5 and create a simple queue instead
const urlsToTestInParallel = 20
const url = 'https://jonassebastianohlsson.com/criticalpathcssgenerator/'

puppeteer.launch().then(browser => {
	for (let i = 0; i < urlsToTestInParallel; i++) {
		penthouse({
			url,
			...otherOptions,
			puppeteer: {
				getBrowser: () => browser
			}
		})
		.then(criticalcss => {
			// finished
		})
	}
})
```

@hummal can you please try this on your machine on `master` in penthouse and see if you get fewer errors?

**Comment by hummal** [comment]
Iam back. Sorry for being late. 
I tried the browser code, but unfortunately puppeteer.launch() always run into an error "kill ERSCH". Dont know why that happen. Maybe there is something strange in my configuration.

**Comment by pocketjoso** [comment]
I see that the example I gave for you to run was not very smart (and it also didn't have the enough config for launching puppeteer safely).

I recently added more documentation about how to run many urls at the same time - can you try that example instead?
https://github.com/pocketjoso/penthouse/blob/master/examples/many-urls.js

**Comment by pocketjoso** [comment]
> On the other hand I can't get the example to run. Dunno why.
`TypeError: Cannot read property 'goto' of undefined`

Hmm. Do any of the examples run for you?

--

> Correct me if Iam wrong, but this is means not threaded. As for my use case it is very important to have the power of all cpu cores to calculate the critical css.

The heaviest work in Penthouse is running Chromium (via Puppeteer) to load pages and prune the critical css (and optionally take screenshots). This is done in one tab per job, and Chromium does take advantage of all your cores. What happens outside of Chromium is less CPU intensive, and if you run enough jobs in parallel you should not see much idle CPU time, at least I don't.

So it depends on your requirements, but it is most likely not very beneficial for you to spawn separate processes. Your different cores are already in use, and starting multiple processes yourself means you will, I think, waste resources having one browser instance per process, instead of sharing it.

**Comment by hummal** [comment]
So testet my original code of my project with the 1.4.1 version of penthouse. It works as expected. 
Changed my code to NOT use child process as explained by you above. Problem: Still the error I fixed in my PR.
`Error: Protocol error (Runtime.callFunctionOn): Cannot find context with specified id undefine`

So I can't test the speed impact of my multi process version. 

Will try on penthouse master with the examples.

**Comment by hummal** [comment]
Basic example is working. But has an error in it. The cssString is 
`color; red` and should be `color: red`

But it is working. So puppeteer.launch is the problem

**Comment by hummal** [comment]
So tested everything again and now its working. Used my windows 10 x64 machine and ubuntu 16.10 x64. 
But I would say in my use case Iam faster with child processes. Because there is about 10 urls per entry to check for one child process. And this in parallel. I can reach much faster execution with 10x10 and my cores are way more used.

Anyway I could use a browser instance for each of the 10 childs.

How to proceed now? The issue still exists and I could fix them with my PR. I need that thing and it is even without child processes a problem.

**Comment by pocketjoso** [comment]
> But has an error in it. The cssString is
> color; red and should be color: red

Thanks, it was just fixed by another pr.

> So tested everything again and now its working

Does this mean you still get your crashes, or not?

>  I can reach much faster execution with 10x10 and my cores are way more used.

@hummal given what I said about Chrome using all your cores, did you try side by side to really verify that you get _much_ faster execution when spawning child processes?

> How to proceed now?

As much as possible, I would like to avoid bigger changes in Penthouse in response to: puppeteer bugs and/or things that only effect a minority of users. As it stands your PR is quite big and I'm not sure how much of it is needed to fix your problems. Could you start splitting it so we can get more clarity?

What is the minimal change required to fix your issue? After we talked a bit more I described why I want the page load logic to work the way it does. And some of your error handling I already merged into master. What remains that makes a difference for you? If you can identify that and open in a _new_ PR, it will be easier for me to merge.

All the commits that had nothing to do with your problem (like fixing/updating deps), please open separate PR's or leave these changes for now. This will also make them easier for me to merge - a PR that just upgrades deps _without breaking tests_ is generally a no-brainer!

Thanks for your patience @hummal!

**Comment by hummal** [comment]
> Does this mean you still get your crashes, or not?
With the master branch of penthouse my error is still present. It is working was just about the examples. 
>@hummal given what I said about Chrome using all your cores, did you try side by side to really verify that you get much faster execution when spawning child processes?
I guess I didn't explained the use case clearly enough.
Lets say I have 24 css files to check. For each of this files I have about 12 sites to check against. 
That means: I use max(CPU CORES) child process, each of them use penthouse for 1 css file with 12 browser tabs ( the test urls). 
If I would left the child process behind I could only test 1 css file with 12 tabs concurrently. So Iam able to have a penhouse process with 12 tabs per process per cpu core. 

But here are the measurement results:
4 Files with each the same amout of testurls (12 Tabs)
ChildProcess: 
1:15
SingleProcess:
2:05

While the CPU usage in single process mode is at 60% avg, it is at 95% avg in child process mode.
Tested with the same code. Only difference is the child fork usage.

>  that only effect a minority of users
I guess it effects every user that uses penthouse to check about 50+ urls. I don't think penthouse is used as a browser plugin that much. Am I that terribly wrong? The most used case is as a nodejs module I think. 

>  I imagine that there are solutions to "avoid navigation.." - the original issue here
That is what they thought. But I proofed it is not. As there are lots of other similar issues which seems to be fixed by setting page.waitFor(500+). It is also adviced by the developers. Because they don't know whats wrong. One guy said he fixed it with going to the newest chrome instead of using chromium puppeteer came with.

> As it stands your PR is quite big and I'm just not sure how much of it is needed to fix your problems. Could you start splitting it so we can get more clarity?
> What is the minimal change required to fix your issue?
The minimal change I guess would be to add page.waitFor(500) before every (2 of them) page.evaluate calls. That should do the trick. 
The other thing I changed was the behaviour the pageloadskiptimeout was calculated. I guess it is in the wrong place. So I moved it into the browser where it belongs.

> All the commits that had nothing to do with your problem (like fixing/updating deps), please open separate PR's or leave these changes for now.
I will follow your advice and open new PR regarding dependecies update, my bugfix for the problem in the minimal version and at least if you agree my moving of pageloadskiptimeout. 
Waiting for your answer for this. :)

> I know I'm not being easy on you. The general needs for Penthouse seem to be a bit different to yours;
You just want the best for penthouse. The same as I do. 
I may be a poweruser of penthouse, but that don't mean we can't optimize it to be a fit for everyone. 
Anyway Iam not a fan of creating another option value for adding the waitFor topic. Because it is indeed a problem for ALL puppeteer users. It could happen every time. And this "workaround" is just in the penthouse code until they fixed their problems. My problem is that I can't wait them to fix it, because it seems this could take ages. So I implement the workaround for it till they fix it. :)

To be honest, we are talking about 500ms. I don't think that anyone will complain about it while he knows that there won't be any broken page requests or crashes because of that. But I admit that it is just my opinion. Yours is the one that counts.

**Comment by hummal** [comment]
PRs are supplied

**Comment by hummal** [comment]
After creating my own library: https://www.npmjs.com/package/crittr I was able to get rid of all the issues penthouse had. Benefits are cleaner code, faster code, more features. 
Thanks for your inspiration.

**Comment by pocketjoso** [comment]
Good job @hummal - long live open source! :) Just started checking it out; I created a PR for a benchmark fix:
https://github.com/hummal/crittr/pull/3
